File: frontend/src/shared/api/ruz.ts
import { z } from 'zod'
import { apiGet } from './http'

const facultySchema = z.object({
  id: z.number(),
  name: z.string(),
  abbr: z.string(),
})

const groupSchema = z.object({
  id: z.number(),
  name: z.string(),
  level: z.number().nullable(),
  type: z.string().nullable(),
  kind: z.number().nullable(),
  spec: z.string(),
  year: z.number().nullable(),
  faculty: facultySchema.nullable(),
})

const teacherSchema = z.object({
  id: z.number(),
  oid: z.number().nullable().optional(),
  full_name: z.string(),
  first_name: z.string().optional(),
  middle_name: z.string().optional(),
  last_name: z.string().optional(),
  grade: z.string().optional(),
  chair: z.string().optional(),
})

const buildingSchema = z.object({
  id: z.number(),
  name: z.string(),
  abbr: z.string(),
  address: z.string(),
})

const auditoriumSchema = z.object({
  id: z.number(),
  name: z.string(),
  building: buildingSchema,
})

const lessonTypeSchema = z.object({
  id: z.number().nullable(),
  name: z.string(),
  abbr: z.string(),
})

const lessonSchema = z.object({
  subject: z.string(),
  subject_short: z.string(),
  type: z.number().nullable(),
  additional_info: z.string(),
  time_start: z.string().nullable(),
  time_end: z.string().nullable(),
  typeObj: lessonTypeSchema.nullable(),
  parity: z.number().nullable(),
  webinar_url: z.string().optional(),
  lms_url: z.string().optional(),
  groups: z.array(groupSchema).optional(),
  teachers: z.array(teacherSchema).optional(),
  auditories: z.array(auditoriumSchema),
})

const daySchema = z.object({
  weekday: z.number(),
  date: z.string(),
  lessons: z.array(lessonSchema),
})

const weekSchema = z.object({
  date_start: z.string(),
  date_end: z.string(),
  is_odd: z.boolean(),
})

const scheduleMetaSchema = z.object({
  source: z.enum(['live', 'cache']),
  is_stale: z.boolean(),
  fetched_at: z.string().nullable().optional(),
  failed_refresh_at: z.string().nullable().optional(),
})

const facultyGroupsSchema = z.object({
  faculty: facultySchema,
  groups: z.array(groupSchema),
})

const groupScheduleSchema = z.object({
  week: weekSchema,
  group: groupSchema,
  days: z.array(daySchema),
  meta: scheduleMetaSchema.nullable().optional(),
})

const teacherScheduleSchema = z.object({
  week: weekSchema,
  teacher: teacherSchema,
  days: z.array(daySchema),
})

export type Faculty = z.infer<typeof facultySchema>
export type Group = z.infer<typeof groupSchema>
export type Teacher = z.infer<typeof teacherSchema>
export type Lesson = z.infer<typeof lessonSchema>
export type GroupSchedule = z.infer<typeof groupScheduleSchema>
export type TeacherSchedule = z.infer<typeof teacherScheduleSchema>

export async function getFaculties(): Promise<Faculty[]> {
  return apiGet('/api/v1/faculties', z.array(facultySchema))
}

export async function getGroupsByFaculty(facultyId: number): Promise<Group[]> {
  const data = await apiGet(`/api/v1/faculties/${facultyId}/groups`, facultyGroupsSchema)
  return data.groups
}

export async function searchGroups(query: string): Promise<Group[]> {
  return apiGet(`/api/v1/groups/search?q=${encodeURIComponent(query)}`, z.array(groupSchema))
}

export async function searchTeachers(query: string): Promise<Teacher[]> {
  return apiGet(`/api/v1/teachers/search?q=${encodeURIComponent(query)}`, z.array(teacherSchema))
}

export async function getGroupSchedule(groupId: number, date: string): Promise<GroupSchedule> {
  return apiGet(`/api/v1/groups/${groupId}/schedule?date=${date}`, groupScheduleSchema)
}

export async function getTeacherSchedule(teacherId: number, date: string): Promise<TeacherSchedule> {
  return apiGet(`/api/v1/teachers/${teacherId}/schedule?date=${date}`, teacherScheduleSchema)
}

File: app/schemas/ruz.py
from datetime import date, datetime, time, timezone
from typing import Literal
from zoneinfo import ZoneInfo

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator

MOSCOW_TZ = ZoneInfo("Europe/Moscow")


def parse_ruz_date(value: date | str) -> date:
    if isinstance(value, date):
        return value

    return date.fromisoformat(value.replace(".", "-"))


def to_utc_datetime(day: date, value: datetime | str | None) -> datetime | None:
    if value in (None, ""):
        return None
    if isinstance(value, datetime):
        if value.tzinfo is None:
            value = value.replace(tzinfo=MOSCOW_TZ)
        return value.astimezone(timezone.utc)

    if "T" in value:
        parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
        if parsed.tzinfo is None:
            parsed = parsed.replace(tzinfo=MOSCOW_TZ)
        return parsed.astimezone(timezone.utc)

    lesson_time = time.fromisoformat(value)
    return datetime.combine(day, lesson_time, tzinfo=MOSCOW_TZ).astimezone(timezone.utc)


class RuzModel(BaseModel):
    model_config = ConfigDict(extra="ignore", populate_by_name=True)


class Faculty(RuzModel):
    id: int
    name: str
    abbr: str = ""


class Group(RuzModel):
    id: int
    name: str
    level: int | None = None
    type: str | None = None
    kind: int | None = None
    spec: str = ""
    year: int | None = None
    faculty: Faculty | None = None


class FacultyGroups(RuzModel):
    faculty: Faculty
    groups: list[Group]


class Teacher(RuzModel):
    id: int
    oid: int | None = None
    full_name: str
    first_name: str = ""
    middle_name: str = ""
    last_name: str = ""
    grade: str = ""
    chair: str = ""


class Building(RuzModel):
    id: int
    name: str
    abbr: str = ""
    address: str = ""


class Auditorium(RuzModel):
    id: int
    name: str
    building: Building


class LessonType(RuzModel):
    id: int | None = None
    name: str = ""
    abbr: str = ""


class Lesson(RuzModel):
    subject: str = ""
    subject_short: str = ""
    type: int | None = None
    additional_info: str = ""
    time_start: datetime | None = None
    time_end: datetime | None = None
    type_obj: LessonType | None = Field(default=None, alias="typeObj")
    parity: int | None = None
    groups: list[Group] = Field(default_factory=list)
    teachers: list[Teacher] = Field(default_factory=list)
    auditories: list[Auditorium] = Field(default_factory=list)
    webinar_url: str = ""
    lms_url: str = ""

    @model_validator(mode="before")
    @classmethod
    def normalize_nullable_lists(cls, data: object) -> object:
        if not isinstance(data, dict):
            return data

        data = data.copy()
        for field in ("groups", "teachers", "auditories"):
            if data.get(field) is None:
                data[field] = []
        return data


class Day(RuzModel):
    weekday: int
    date: date
    lessons: list[Lesson] = Field(default_factory=list)

    @model_validator(mode="before")
    @classmethod
    def normalize_lesson_times(cls, data: object) -> object:
        if not isinstance(data, dict):
            return data

        day_date = parse_ruz_date(data["date"])
        data = data.copy()
        data["date"] = day_date
        lessons = data.get("lessons") or []
        data["lessons"] = [
            {
                **lesson,
                "time_start": to_utc_datetime(day_date, lesson.get("time_start")),
                "time_end": to_utc_datetime(day_date, lesson.get("time_end")),
            }
            if isinstance(lesson, dict)
            else lesson
            for lesson in lessons
        ]
        return data


class Week(RuzModel):
    date_start: date
    date_end: date
    is_odd: bool

    @field_validator("date_start", "date_end", mode="before")
    @classmethod
    def normalize_date(cls, value: date | str) -> date:
        return parse_ruz_date(value)


class ScheduleMeta(RuzModel):
    source: Literal["live", "cache"]
    is_stale: bool = False
    fetched_at: datetime | None = None
    failed_refresh_at: datetime | None = None


class GroupSchedule(RuzModel):
    week: Week
    group: Group
    days: list[Day] = Field(default_factory=list)
    meta: ScheduleMeta | None = None


class TeacherSchedule(RuzModel):
    week: Week
    teacher: Teacher
    days: list[Day] = Field(default_factory=list)
